www/js/app.js

```diff
@@ -2303,6 +2318,112 @@ function articleLoadedSW (iframeArticleContent) {
+    // (having this condition prevents very erratic popover placement in IE11, for example, so the feature is disabled for such browsers)
+    if (appstate.wikimediaZimLoaded && params.showPopoverPreviews && 'matches' in Element.prototype) {
+        const iframeDoc = win.document;
+        if (iframeDoc) {
+            // Attach the popover CSS to the current article document
+            uiUtil.attachKiwixPopoverCss(iframeDoc);
+            // Add event listeners to iframe window to check for links
+            win.addEventListener('mouseover', handlePopoverEvents, true);
+            win.addEventListener('focus', handlePopoverEvents, true);
+            // Conditionally add event listeners to support touch events with fallback to pointer events
+            if (window.navigator.maxTouchPoints > 0) {
+                win.addEventListener('touchstart', handlePopoverEvents, true);
+            } else {
+                win.addEventListener('pointerdown', handlePopoverEvents, true);
+            }
+        }
+    }
+}
+
+/**
+ * Event handler for attaching preview popovers
+ * @param {Event} event The event produced by the mouseover or focus action
+ */
+function handlePopoverEvents (event) {
+    // Check if the hovered or focused element or its parent is a link (but throttle its activation)
+    if (window.popoverThrottle) return;
+    window.popoverThrottle = true;
+    setTimeout(function () {
+        let a = event.target;
+        const iframeDoc = a.ownerDocument;
+        if (iframeDoc) {
+            const iframeWindow = iframeDoc.defaultView;
+            while (a && a !== iframeWindow && a.nodeName !== 'A') {
+                a = a.parentNode;
+            }
+            // If a link was hovered or focused, process it
+            if (a && a.nodeName === 'A') {
+                // console.debug(event.type, event.target, a);
+                const suppressContextMenuHandler = function (e) {
+                    e.preventDefault();
+                    e.stopPropagation();
+                };
+                // Prevent context menu on this anchor element
+                a.addEventListener('contextmenu', suppressContextMenuHandler, true);
+                if (/touchstart|pointerdown/.test(event.type)) {
+                    a.touched = true; // Used to prevent dismissal of popver on mouseout if initiated by touch
+                }
+                if (a.style.userSelect === undefined) {
+                    // This prevents selection of the text in a touched link in Safari for iOS and Edge Legacy / UWP
+                    a.style.webkitUserSelect = 'none';
+                    a.style.msUserSelect = 'none';
+                }
+                // Check if a popover div is currently being hovered
+                const divs = iframeDoc.getElementsByClassName('kiwixtooltip');
+                let divIsHovered = false;
+                Array.prototype.slice.call(divs).forEach(function (div) {
```

Comment:
Why do you have to invoke `Array.prototype.slice` statically? Is this for pollyfills? You can't just do `divs.forEach()`?

Reply:
Old habit. I can just use `Array.from()` because we have that polyfilled. Before the polyfill, we always used this formula in order to support older browsers that cannot do `forEach` over a Node list. So it became second nature... Will simplify.